src/kernel/arch/x86/paging.zig

```diff
@@ -0,0 +1,122 @@
+}
+
+///
+/// Map a page directory. The addresses passed must be 4KB aligned and be the same distance apart.
+///
+/// Arguments:
+///     IN entry: *Directory - The directory to map
+///     IN phys_start: u32 - The physical address at which to start mapping
+///     IN phys_end: u32 - The physical address at which to stop mapping
+///     IN virt_start: u32 - The virtual address at which to start mapping
+///     IN virt_end: u32 - The virtual address at which to stop mapping
+///     IN allocator: *Allocator - The allocator to use to map any tables needed
+///
+fn map_dir(dir: *Directory, phys_start: usize, phys_end: usize, virt_start: usize, virt_end: usize, allocator: *std.mem.Allocator) void {
+    assert(phys_start <= phys_end);
+    assert(phys_end - phys_start == virt_end - virt_start);
+    assert(isAligned(phys_start, PAGE_SIZE));
+    assert(isAligned(phys_end, PAGE_SIZE));
+
+    var virt_addr = virt_start;
+    var phys_addr = phys_start;
+    var page = virt_addr / PAGE_SIZE_4KB;
+    var entry_idx = virt_addr / PAGE_SIZE;
+    while (entry_idx < ENTRIES_PER_DIRECTORY and virt_addr < virt_end) {
+        map_dir_entry(&dir.entries[entry_idx], virt_addr, phys_addr, allocator);
+        phys_addr += PAGE_SIZE;
+        virt_addr += PAGE_SIZE;
+        entry_idx += 1;
+    }
+}
+
+// These two functions were added to Zig master since v0.4 but using master is blocked on https://github.com/ziglang/zig/issues/2616
+// so they are copied here.
+// TODO: Remove and replace with stdlib functions once #2616 is fixed.
+fn alignBackward(addr: usize, alignment: usize) usize {
+    return addr & ~(alignment - 1);
+}
+
+fn isAligned(addr: usize, alignment: usize) bool {
+    return alignBackward(addr, alignment) == addr;
+}
+
+fn page_fault(state: *arch.InterruptContext) void {
+    @panic("Page fault");
+}
+
+///
+/// Initialise x86 paging, overwriting any previous paging set up.
+///
+/// Arguments:
+///     IN mem_profile: *const MemProfile - The memory profile of the system and kernel
+///     IN allocator: *std.mem.Allocator - The allocator to use
+///
+pub fn init(mem_profile: *const MemProfile, allocator: *std.mem.Allocator) void {
+    // Calculate start and end of mapping
+    const v_start = alignBackward(@ptrToInt(mem_profile.vaddr_start), PAGE_SIZE);
+    const v_end = std.mem.alignForward(@ptrToInt(mem_profile.vaddr_end) + mem_profile.fixed_alloc_size, PAGE_SIZE);
+    const p_start = alignBackward(@ptrToInt(mem_profile.physaddr_start), PAGE_SIZE);
+    const p_end = std.mem.alignForward(@ptrToInt(mem_profile.physaddr_end) + mem_profile.fixed_alloc_size, PAGE_SIZE);
+
+    var tmp = allocator.alignedAlloc(Directory, PAGE_SIZE_4KB, 1) catch unreachable;
+    var kernel_directory = @ptrCast(*Directory, tmp.ptr);
+    @memset(@ptrCast([*]u8, kernel_directory), 0, @sizeOf(Directory));
+
+    map_dir(kernel_directory, p_start, p_end, v_start, v_end, allocator);
+    const dir_physaddr = @ptrToInt(kernel_directory) - 0xC0000000;
```

Comment:
Could 0xC0000000 become a global const as it is the virtual location of the kernel